core/src/Streamly/Internal/Data/Stream/Common.hs

```diff
@@ -1,39 +1,83 @@
-------------------------------------------------------------------------------
--- Conversion to and from direct style stream
-------------------------------------------------------------------------------
+{-# LANGUAGE UndecidableInstances #-}
+{-# OPTIONS_GHC -Wno-orphans #-}
 
-#ifdef USE_IS_STREAM
-#define IS_STREAM IsStream t,
-#define FROM_STREAM fromStream
-#define TO_STREAM toStream
-#define STREAM t
+-- |
+-- Module      : Streamly.Internal.Data.Stream.Common
+-- Copyright   : (c) 2017 Composewell Technologies
+-- License     : BSD-3-Clause
+-- Maintainer  : streamly@composewell.com
+-- Stability   : experimental
+-- Portability : GHC
+--
+-- Low level functions using StreamK as the intermediate stream type. These
+-- functions are used in SerialT/AsyncT/AheadT/ParallelT stream modules to
+-- implement their instances..
+--
+module Streamly.Internal.Data.Stream.Common
+    (
+    -- * Running Effects
+      drain
+
+    -- * Conversion operations
+    , fromList
+    , toList
+
+    -- * Fold operations
+    , foldrM
+    , foldrMx
+    , foldr
+
+    , foldlx'
+    , foldlMx'
+    , foldl'
+    , fold
+
+    -- * Zip style operations
+    , eqBy
+    , cmpBy
+    )
+where
+
+#include "inline.hs"
+
+import Streamly.Internal.Data.Fold.Type (Fold (..))
+
+#ifdef USE_STREAMK_ONLY
+import qualified Streamly.Internal.Data.Stream.StreamK as S
+import qualified Streamly.Internal.Data.Stream.StreamK.Type as S
 #else
-#define IS_STREAM
-#define FROM_STREAM id
-#define TO_STREAM id
-#define STREAM K.Stream
+import qualified Streamly.Internal.Data.Stream.StreamD.Type as S
 #endif
 
+import qualified Streamly.Internal.Data.Stream.StreamK.Type as K
+import qualified Streamly.Internal.Data.Stream.StreamD.Type as D
+
+import Prelude hiding (foldr, repeat)
+
+------------------------------------------------------------------------------
+-- Conversion to and from direct style stream
+------------------------------------------------------------------------------
+
 -- These definitions are dependent on what is imported as S
 {-# INLINE fromStreamS #-}
-fromStreamS :: (IS_STREAM Monad m) => S.Stream m a -> STREAM m a
-fromStreamS = FROM_STREAM . S.toStreamK
+fromStreamS :: Monad m => S.Stream m a -> K.Stream m a
+fromStreamS = S.toStreamK
 
 {-# INLINE toStreamS #-}
-toStreamS :: (IS_STREAM Monad m) => STREAM m a -> S.Stream m a
-toStreamS = S.fromStreamK . TO_STREAM
+toStreamS :: Monad m => K.Stream m a -> S.Stream m a
+toStreamS = S.fromStreamK
 
 {-# INLINE toStreamD #-}
-toStreamD :: (IS_STREAM Monad m) => STREAM m a -> D.Stream m a
-toStreamD = D.fromStreamK . TO_STREAM
+toStreamD :: Monad m => K.Stream m a -> D.Stream m a
+toStreamD = D.fromStreamK
```

Comment:
We can remove toStreamS/fromSrteamS/toStreamD, and directly use toStreamK/fromSrteamK?

Reply:
Yes